File: DENN Verification 2/de-msv-2.c
#include<stdio.h>
#include<time.h>
#include<stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdbool.h>
#include <math.h>

#define ITERATIVE_PARAM_PATH        "../bpnn/dataset/iterative_param.txt"

#define D   IN_N
#define Q   HIDDEN_N
#define L   OUT_N
#define U1  LEARN_RATE1
#define U2  LEARN_RATE2

#define SEEK_SET   0
#define RAND_MAX   32767

struct _iobuf {
	char *_ptr;
	int   _cnt;
	char *_base;
	int   _flag;
	int   _file;
	int   _charbuf;
	int   _bufsiz;
	char *_tmpfname;
	};
typedef struct _iobuf FILE;

FILE *in_file;  // ITERATIVE_PARAM

int main(void)
{
    printf("the program is running!\n");

    in_file = fopen(ITERATIVE_PARAM_PATH, "r");

    double *res = (double *)malloc(sizeof(double) * 10);

    if (!in_file) {
        fprintf(stderr, "Cannot open %s!", in_file);
    }

    char buffer[256];
    int index = 0;

    while (fgets(buffer, 256, in_file) != NULL) {
        char *token = strtok(buffer, " ");
        for (int i = 0; i < 27; i ++) {
            if (i == 25) {
                res[index] = strtod(token, NULL);
                token = strtok(NULL, " ");
            }
            strtod(token, NULL);
            token = strtok(NULL, " ");
        }
        index ++;
        if (index == 10) {
            break;
        }
    }

    for (int i = 0; i < 10; i ++) {
        printf("%f\n", res[i]);
    }

    /* 对网络结构优化过程的验证

    define p: res[0] >= res[1];
    define q: res[1] >= res[2];
    alw (p and q)

    结论：在网络结构的优化过程中，种群中的网络结构总是变得越来越好。
    */
    fclose(in_file);

	return 0;
}

File: DENN Verification 2/de-msv-2.m
frame(in_file, main_res) and (
struct _iobuf {
char *_ptr and 
int _cnt and 
char *_base and 
int _flag and 
int _file and 
int _charbuf and 
int _bufsiz and 
char *_tmpfname 
};
FILE *in_file and skip;
float main_res[10] and skip;
 function main ( int  RValue )
 {
     frame(main_buffer,main_index,main_2_temp$_1,main_2_token,main_2_i,main_i,return,break$) and (
     int break$<==0 and skip;
     int return<==0 and skip;
     output ("the program is running!\n") and skip;
     in_file:=fopen("iterative_param.txt","r");
     if(!in_file) then 
     {
         fprintf(stderr,"Cannot open %s!",in_file) and skip
         
     }
     else 
     {
          skip 
     };
     char main_buffer[256] and skip;
     int main_index<==0 and skip;
     break$<==0 and skip;
     char* main_2_temp$_1 and skip;
     main_2_temp$_1:=fgets(main_buffer,256,in_file);
     while( break$=0 AND   (main_2_temp$_1!=NULL) )
     {
         char* main_2_token and skip;
         main_2_token:=strtok(main_buffer," ",RValue);
         int main_2_i<==0 and skip;
         
         while( (main_2_i<27) )
         {
             if(main_2_i=26) then 
             {
                 main_res[main_index]:=strtod(main_2_token,NULL,RValue);
                 main_2_token:=strtok(NULL," ",RValue)
             }
             else 
             { 
			      strtod(main_2_token,NULL,RValue) and skip;
                  main_2_token:=strtok(NULL," ",RValue);
                  skip 
             };
             main_2_i:=main_2_i+1
         };
         main_index:=main_index+1;
         if(main_index=10) then 
         {
             break$<==1 and skip
             
         }
         else 
         {
		      main_2_temp$_1:=fgets(main_buffer,256,in_file);
              skip 
         }
     };
     break$<==0 and skip;
     int main_i<==0 and skip;
     
     while( (main_i<10) )
     {
         output (main_i, " ", main_res[main_i],"\n") and skip;
         main_i:=main_i+1
         
     };
     fclose(in_file) and skip;
     return<==1 and RValue:=0;
     skip
     )
 };
  main(RValue)
 )
